### PR TITLE
feat(moe): mxfp4-resident MoE experts for DeepSeek-V4-Flash LoRA

### PR DESCRIPTION
**Pull request by excepshenal**
# What does this PR do ?

**What**
Adds an mxfp4-resident storage path for the base experts of DeepSeek-V4-Flash during LoRA training. Experts are kept packed as fp4-e2m1 + e8m0 block scales (the format they already ship in) at steady state and dequantized on the fly inside the grouped-GEMM forward/backward, instead of being materialized in bf16. This lets DSV4-Flash LoRA fit and run on a single 8×H100 node.

**Why**
The experts are ~90% of DSV4-Flash's parameters (~277B → ~69 GiB/rank bf16 at ep_size=8). Dequantizing them to bf16 at load both balloons memory and makes load slow. Keeping them in their native mxfp4 form — only legal because LoRA freezes the base (no master copy / optimizer state needed for experts) — removes both costs.

**How it works**
- Packed-resident storage. quantization/mxfp4.py provides pack/unpack + MXFP4GroupedMM, a grouped-GEMM autograd Function that saves only the packed weights and re-dequantizes in backward (no bf16 weight kept alive; no weight grad).
- Passthrough load (no bf16 ever). The DSV4 state-dict adapter gains expert_storage_format="mxfp4": from_hf aggregates per-expert packed int8 + e8m0 scales straight into *_packed/*_scales params, and to_hf splits them back for the DCP planner. Packing is orthogonal to the expert-stack and gate‖up concat (it's along the contraction dim), so no unpacking is needed at load. Modules are built packed-at-init so the checkpoint loads directly into packed params.
- EP integration. ExpertParallel shards the packed params on the expert dim; works on the torch_mm GroupedExperts path with dispatcher: torch.

**Validation (8×H200, full 43-layer DSV4-Flash, ep_size=8, tilelang)**
- Numerical parity: step-0 LoRA loss matches a bf16-experts baseline (bit-identical at short seq; within bf16 noise at 4k packed seq). Unit tests assert forward/backward equivalence vs bf16 and bit-exact decode vs the existing fp4 dequant, incl. a real-checkpoint-gated test.
- Memory: ~38 GiB/rank (mxfp4) vs ~84 GiB/rank (bf16) — ~2.2× less.
- Load time: ~63 s (mxfp4) vs ~16.5 min (bf16, warm cache) — passthrough skips the load-time dequant+materialize.
- Throughput (4096-token packed seq): ~6,567 tps (mxfp4) vs ~7,740 tps (bf16) — mxfp4 ~1.18× slower; the dequant overhead shrinks with longer seq (was ~2.2× at short seq) since dequant is a fixed cost amortized over a larger GEMM.

**Known limitations / scope**
- EP required. mxfp4 needs the torch_mm GroupedExperts path with ep_size > 1; single-GPU is guarded with a clear error. DeepEP experts are not yet supported (the conversion skips them with a warning).
ep_shard > 1 (multi-node where experts are FSDP-wrapped) still hits a loud FSDP-on-int8 error — needs packed experts excluded from FSDP before that config; single-node ep_size=8/ep_shard=1 avoids it.
- Dequant is an unfused pure-torch kernel; further speedups (dequant-once-per-grad-accum, side-stream overlap, fused/Blackwell-native GEMM) are follow-ups.

# Changelog
- New: expert_storage_format-driven mxfp4 storage for base MoE experts (quantization/mxfp4.py, moe/quantized_experts.py, _peft/lora* for LoRA-targeted experts).
- New: DSV4 adapter mxfp4 passthrough load/save (no bf16 expert materialization).
- New: example config examples/llm_finetune/deepseek_v4/deepseek_v4_flash_hellaswag_lora_mxfp4.yaml (8-GPU, ep_size=8, tilelang, mxfp4).
- New PEFT config field peft.expert_weight_format: "bf16" | "mxfp4" (default bf16; mxfp4 requires experts: torch_mm).
- Fix: skip random init for packed experts; set requires_grad at nn.Parameter construction in ExpertParallel (int8 EP-sharding blocker); guard single-GPU (mxfp4 requires EP).
- Fix (general): clean process exit (destroy_process_group) and deadlock-safe distributed validation (lockstep across ranks) in recipes/llm/train_ft.py.
- Perf: drop a redundant .contiguous() in the expert grouped GEMM (cuBLAS consumes the transposed view).

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?

### CODE REVIEW

**Comment by BitPhinix** [comment]
We're using this, would be nice to get this merged!

**Comment by HuiyingLi** [comment]
Thank you! Looking at it now!

**Comment by HuiyingLi** [comment]
Could you please merge main&resolve conflict, thanks

File: nemo_automodel/components/moe/layers.py [hunk 1]
```diff
@@ -449,7 +449,11 @@ def _route_scores(self, scores: torch.Tensor) -> tuple[torch.Tensor, torch.Tenso
             weights = original_scores.gather(1, indices)
         elif self.score_func == "sqrtsoftplus":
             # sqrt(softplus(x)) = sqrt(log(1 + exp(x))), used in DeepSeek V4.
-            scores = torch.sqrt(F.softplus(scores.float()))
+            # clamp_min keeps the sqrt argument strictly positive: softplus(x) underflows
+            # to exactly 0.0 in fp32 for very negative logits (x <~ -104), and sqrt'(0) = inf
+            # makes the backward NaN. The clamp bounds the gradient with a negligible
+            # (sqrt(1e-12) = 1e-6) change to the forward value.
+            scores = torch.sqrt(F.softplus(scores.float()).clamp_min(1e-12))
```
Comment (HuiyingLi) [inline nemo_automodel/components/moe/layers.py:455]: Looks like DSV4 hf impl doesn't have this clamp. Could you please explaining reasons for adding it here, did you need it during your training? Thanks!

**Comment by HuiyingLi** [comment]
Hi! Would you mind if I pushed some changes directly to your PR branch? I’d like to help with a few refinements. Thanks!

**Comment by BitPhinix** [comment]
Anything blocking to get this merged?

**Comment by akoumpa** [comment]
Hi @excepshenal + @BitPhinix , unfotunately, i see two failures (e.g., https://github.com/NVIDIA-NeMo/Automodel/actions/runs/32448166385/job/97083876807?pr=2548 https://github.com/NVIDIA-NeMo/Automodel/actions/runs/32448166385/job/97083876829?pr=2548). 

If you can fix these + merge main, then i can help rerun ci

**Comment by HuiyingLi** [comment]
Thanks for the reminder @BitPhinix  @akoumpa  will work on merging this

**Comment by HuiyingLi** [comment]
Validated **100 optimizer steps with HybridEP** on PR head `f434b61e3c697cda4fc397d6f092703f03b4804e`.

[W&B run](https://wandb.ai/Nemo-automodel/automodel-dsv4/runs/p2548f434j18526759) — complete training and validation curves; run files include the exact config, command, source fingerprints, JSONL metrics, and result summary.

- **Hardware:** 1 node × 8 H100 80GB, DFW interactive job `18526759` (`pool0-01266`).
- **Model/backend:** full 43-layer DeepSeek-V4-Flash pretrained model; FSDP2 + EP8, TP = CP = PP = 1; HybridEP token dispatch, `torch_mm` MXFP4 routed experts, TileLang attention, activation checkpointing.
- **Training:** MaskedCrossEntropy; BF16 attention/routed/shared expert LoRA (rank 8, alpha 32); TE FusedAdam with FP32 master weights/moments, LR `1e-5`, seed 1234.
- **Data/batches:** full HellaSwag train split, packed sequence length 4096, configured global batch size 128, microbatch size 1 per GPU, gradient accumulation 16. `num_epochs=null` lets the scheduler run enough epochs to reach 100 steps; native smaller epoch-end batches are retained.
- **Validation:** full HellaSwag validation every 25 steps, using the standard distributed sampler/drop_last (309,613 supervised tokens per validation pass).

**Result:** 100/100 optimizer steps and four validation passes completed, followed by normal shutdown (exit code 0). All 100 training losses and gradient norms are finite.

| Metric | Result |
| --- | --- |
| First → last train loss | 2.359608 → 1.522959 |
| Mean first 10 → last 10 train loss | 2.287156 → 1.536417 |
| Peak training allocated memory | 51.23 GiB on rank 0 (recipe metric) |
| Launch-to-exit elapsed time | 115.5 min |

| Completed optimizer steps | Validation loss |
| --- | --- |
| 25 | 1.990157 |
| 50 | 1.795143 |
| 75 | 1.672393 |
| 100 | 1.530364 |

The checkout stayed clean and all 724 tracked source-file fingerprints matched the launch manifest after completion. This extends the earlier 1-step full-model test; checkpoint saving remains disabled. This run covers single-node EP8.

**Comment by HuiyingLi** [comment]
Completed Qwen3 MoE regression validation for head `f434b61e3c697cda4fc397d6f092703f03b4804e`. **All recorded numerical comparisons matched exactly, and peak training allocated memory was unchanged in the tested configurations.**

The reason for this test is that the communication refactor also changes the implementation used by existing ordinary BF16 experts and BF16 LoRA experts. The earlier variable-token-count LoRA fix in [#3631](https://github.com/NVIDIA-NeMo/Automodel/pull/3631) duplicated the parent expert communication logic; this refactor extracts shared helpers for the ordinary, LoRA, and MXFP4 variants. We therefore need to verify that sharing this code preserves existing model behavior, including unequal token counts across EP ranks and activation-checkpoint recomputation. Torch EP exercises `_gather_ep_inputs` / `_combine_ep_output`; DeepEP and HybridEP exercise `_dispatch_tokens`. All tests explicitly use `experts: torch_mm` to exercise the affected classes.

Versions compared, separately within each dispatcher:

- Before refactor: `a3a71cf1fb1bf15c7ee9413118b1647eecd8462e`.
- After refactor: `f434b61e3c697cda4fc397d6f092703f03b4804e`.
- Fixed PR base: `48540bbe799a10829f4be4b139e2b6971c5ce74a`.
- Each dispatcher ran before, a repeat of before, after, and the fixed base, with identical initialization and inputs.

**Small full-model E2E:** real two-layer Qwen3 MoE with native EP2/FSDP2, ordinary BF16 weights, forward → CE → backward → AdamW. Matrix: Torch / DeepEP / HybridEP × ordinary training / BF16 LoRA × per-rank token counts 256/256 or 256/512 × activation checkpointing off/on. Each configuration ran three optimizer steps. Hooks asserted the actual expert-entry shapes on both ranks, including recomputation. The LoRA cases used deterministic nonzero adapters to exercise both A and B gradients from the first step.

All **96 runs** passed. **38,016 tensor comparisons** were identical, covering logits, loss, input-activation gradients, trainable-parameter gradients, updates, and resulting parameters.

**Full pretrained Qwen3-30B-A3B LoRA:** 48 layers, checkpoint revision `ad44e777bcd18fa416d9da3bd8f70d33ebb85d39`, BF16 LoRA rank 32 / alpha 32 / `target_modules: ["*"]`, EP8/FSDP2, activation checkpointing enabled, TE attention with RoPE fusion disabled, and MaskedCrossEntropy. Based on `qwen3_moe_30b_lora.yaml`, with GBS16, local batch2, accumulation1, seed42, and the recipe's AdamW settings (LR 1e-3). Checkpoint writing was disabled.

Torch and HybridEP each ran all four versions/controls for **10 optimizer steps plus validation**. The full HellaSwag train split (39,905 records) and validation split (10,042 records) were loaded without sample limits; training used ten steps from the shuffled full split. Validation retained native sampler/drop-last behavior and evaluated 313,901 supervised tokens.

Every step's input hashes matched. Training loss, global gradient norm, and validation loss matched exactly within each dispatcher. All **554,880 per-parameter gradient/update hash comparisons** passed; **166,464 full-tensor comparisons** also matched, covering initial/final adapters and complete gradient/update tensors at steps 0 and 9.

| Dispatcher | Train loss, first → last | Validation loss | Peak training allocated GiB, before → after |
|---|---:|---:|---:|
| Torch | 2.83743262 → 2.04928637 | 2.04133296 | 11.841555 → 11.841555 |
| HybridEP | 2.82992220 → 2.04484391 | 2.04406595 | 11.851455 → 11.851455 |

Memory above is the maximum PyTorch training allocation across all eight ranks. Validation and initialization can peak higher.

**Native performance control:** the instrumented HybridEP runs showed about a 4% median timing difference. Subtracting local observation time can retain waits on another rank's gradient/tensor collection, so a separate control used the unmodified `train_ft.py` entry point with observation hooks and tensor dumps absent. Order: before → after → after → before, 30 steps each, validation disabled, discarding steps 0–4. Step duration was derived from the native token-count/TPS metrics.

| Run order | Steady median seconds | Steady mean seconds |
|---|---:|---:|
| Before 1 | 0.876497 | 0.932579 |
| After 1 | 0.874532 | 0.950197 |
| After 2 | 0.878155 | 0.968574 |
| Before 2 | 0.868624 | 0.930698 |

Pooled steady median: **0.874355s before → 0.877377s after (+0.35%)**. All 30-step losses, gradient norms, and token counts matched across all four runs. These are short performance controls; the table includes the means to expose timing variability.

All work ran in DFW interactive job `18537759`, `pool0-01618`, 8×H100 80GB, using `nemo-automodel-nightly-20260728.sqsh` (PyTorch 2.13.0a0 NVIDIA build, CUDA 13.3, Transformers 5.12.1). Every run exited successfully; the allocation completed with exit code 0 and was released. The PR worktree remained clean.

Internal DFW artifacts, including the report, exact configs/commands, immutable source snapshots and hashes, native logs, comparison scripts, and tensor snapshots:

```text
/lustre/fsw/portfolios/coreai/users/huiyingl/dsv4/Automodel-pr-2548-runs/qwen-regression-20260914/job18537759/
```

Start with `README.md`, `result.json`, `small-comparison.json`, `full-comparison.json`, and `performance-comparison.json`. This validates the tested Qwen3 MoE communication paths; coverage of other models/topologies and long-run convergence remains separate.
